devito/petsc/solve.py

```diff
@@ -83,38 +85,77 @@ def separate_eqn(eqn, target):
     where F(target) = b.
     """
     zeroed_eqn = Eq(eqn.lhs - eqn.rhs, 0)
-    tmp = eval_time_derivatives(zeroed_eqn.lhs)
-    b, F_target = remove_target(tmp, target)
-    return -b, F_target
+    zeroed_eqn = eval_time_derivatives(zeroed_eqn.lhs)
+    target_funcs = set(generate_targets(zeroed_eqn, target))
+    b, F_target = remove_target(zeroed_eqn, target_funcs)
+    return -b, F_target, target_funcs
+
+
+def generate_targets(eq, target):
+    """
+    Extract all the functions that share the same time index as the target,
+    but may have different spatial indices.
+    """
+    funcs = retrieve_functions(eq)
+    if any(dim.is_Time for dim in target.dimensions):
```

Comment:
aka `isinstance(target, TimeFunction)`?
and then in the `then` part of the branch you could
```
time_dim = target.time_dim
```

Reply:
But what if the target is `u.forward`, i.e., `u(t + dt, x, y)`? In this case, I need to extract `time_idx = t + dt` specifically, as I only want to retrieve the `funcs` that correspond to this time index, and not those with a different time index, such as `t`.

Reply:
the indices should be a DimensionTuple so you should be able to just do

```
if isinstance(target, TimeFunction):
   time_idx = target.indices[target.time_dim]
```

i.e 

```
In [7]: f = TimeFunction(name="f", grid=grid, space_order=8)

In [8]: f
Out[8]: f(t, x)

In [9]: f.indices[f.time_dim]
Out[9]: t
```

Reply:
Oh yes, great thanks